doc/configuration.md
# Configuration

This document describes how you can configure your GDK environment.

## Git configuration

Git has features which are disabled by default, and would be great to enable to
be more effective with Git. Run `rake git:configure` to set the recommendations
for some repositories within the GDK.

To set the configuration globally, run `rake git:configure[true]`. When using
`zsh`, don't forget to escape the square brackets: `rake git:configure\[true\]`.

## GDK configuration

There are many configuration options for GDK. GDK can be configured using either:

- [`gdk.yml`](#gdkyml) configuration file.
- [Loose files](#loose-files-deprecated) (deprecated).

### `gdk.yml`

Placing your settings in `gdk.yml` at the root of GDK is the preferred configuration
method.

To see available configuration settings, see [`gdk.example.yml`](../gdk.example.yml).

This file contains all possible settings with example values. Note
that these values may not be the default that GDK will use.

If you want to check which settings are in place, you can run `rake dump_config`, which will print
all applied settings in a YAML structure.

#### Overwriting configuration files

Any configuration file managed by GDK will be overwritten
whenever there are changes in its source (a `.example` or `.erb`
file). When GDK overwrites a configuration file it moves the original file
into the `.backups` subdirectory of your GDK installation.

If you have local changes that you don't want GDK to touch you can
protect individual configuration files. For example:

```yaml
# in gdk.yml
gdk:
  protected_config_files:
  - 'gitaly/*.toml'
```

> Note that `gdk.yml` is not managed by GDK, and GDK will
never overwrite it.

#### Notable settings

Here are a few settings worth mentioning:

| Setting                | Default | Description                                                                                |
|------------------------|---------|--------------------------------------------------------------------------------------------|
| `port`                 | `3000`  | Select the port to run GDK on, useful when running multiple GDKs in parallel.              |
| `webpack.port`         | `3808`  | Also useful to configure when running GDKs in parallel. [See below for more webpack options](#webpack-settings) |
| `gitlab_pages.port`    | `3010`  | Specify on which port GitLab Pages should run. See also the [Pages guide](howto/pages.md). |
| `relative_url_root`    | `/`     | When you want to test GitLab being available on a different path than `/`, e.g. `/gitlab`. |
| `object_store.enabled` | `false` | Set this to `true` to enable Object Storage with MinIO.                                    |
| `registry.enabled`     | `false` | Set this to `true` to enable container registry.                                           |
| `geo.enabled`          | `false` | Set this to `true` to enable Geo (for now it just enables `postgresql-geo` and `geo-cursor` services). |

#### GDK settings

There are also a few settings that configure the behavior of GDK itself:

| Setting                           | Default | Description                                                                                      |
|-----------------------------------|---------|--------------------------------------------------------------------------------------------------|
| `gdk.ask_to_restart_after_update` | `true`  | Set this to `false` if you do not wish to be prompted to restart your GDK after an update. |
| `gdk.debug`                       | `false` | Set this to `true` to enable increased output. |
| `gdk.overwrite_changes`           | `false` | When set to `true` `gdk reconfigure` will overwrite files and move the old version to `.backups`.|
| `gdk.protected_config_files`           | `[]` | Contains file names / globs of configuration files GDK should not overwrite. |

##### Experimental GDK settings

Experimental settings may be promoted to stable settings or they may be deprecated.

| Setting | Default | Description |
|---------|---------|-------------|
| `gdk.experimental.quiet` | `false` | Set this to `true` to reduce the level of output when updating the GDK. |
| `gdk.experimental.auto_reconfigure` | `false` | Set this to `true` to automatically run a `gdk reconfigure` after a successful `gdk update`. |

### NGINX settings

| Setting | Default | Description |
|---------|---------|-------------|
| `nginx.enabled` | `false` | Set this to `true` to enable the `nginx` service. |
| `nginx.listen` | `127.0.0.1` | Set this to be the IP that NGINX will listen on. |
| `nginx.bin` | `/usr/sbin/nginx` | Set this to the path to your `nginx` binary. |
| `nginx.ssl.certificate` | `localhost.crt` | This maps to [NGINX's `ssl_certificate`](https://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_certificate). |
| `nginx.ssl.key` | `localhost.key` | This maps to [NGINX's `ssl_certificate_key`](https://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_certificate_key). |
| `nginx.http2.enabled` | `false` | Set this to `true` to enable HTTP/2 support. |

See [configuring NGINX](howto/nginx.md) for a comprehensive guide.

### Loose files (deprecated)

Before `gdk.yml` was introduced, GDK could be configured through a
bunch of loose files, where each file sets one setting.

It is still possible to use these loose files, but it's deprecated and
will be removed in the future. A migration path will be provided
when this option is removed.

Below is a table of all the settings that can be set this way:

| Filename                     | Type         | Default                                                                              |
|------------------------------|--------------|--------------------------------------------------------------------------------------|
| `host` / `hostname`          | string or IP | `127.0.0.1`                                                                          |
| `port`                       | number       | `3000`                                                                               |
| `https_enabled`              | boolean      | `false`                                                                              |
| `relative_url_root`          | string       | `/`                                                                                  |
| `webpack_host`               | string or IP | `127.0.0.1`                                                                          |
| `webpack_port`               | number       | `3808`                                                                               |
| `registry_enabled`           | boolean      | `false`                                                                              |
| `registry_port`              | number       | `5000`                                                                               |
| `registry_image`             | string       | `registry.gitlab.com/gitlab-org/build/cng/gitlab-container-registry:v2.9.1-gitlab`   |
| `object_store_enabled`       | boolean      | `false`                                                                              |
| `object_store_port`          | number       | `9000`                                                                               |
| `postgresql_port`            | number       | `5432`                                                                               |
| `postgresql_geo_port`        | number       | `5432`                                                                               |
| `gitlab_pages_port`          | number       | `3010`                                                                               |
| `auto_devops_enabled`        | boolean      | `false`                                                                              |
| `auto_devops_gitlab_port`    | number       | `rand(20000..24999)`                                                                 |
| `auto_devops_registry_port`  | number       | `auto_devops_gitlab_port + 5000`                                                     |
| `google_oauth_client_secret` | ?            | ?                                                                                    |
| `google_oauth_client_id`     | ?            | ?                                                                                    |

### Configuration precedence

GDK will use the following order of precedence when selecting the
configuration method to use:

- `gdk.yml`
- Loose file
- Default value

### Reading the configuration

To print settings from the configuration you can use `gdk config get <setting>`.

More information on the available `gdk` commands is found in
[GDK commands](gdk_commands.md#configuration).

### Implementation detail

Here are some details on how the configuration management is built.

#### GDK::ConfigSettings

This is the base class and the engine behind the configuration
management. It defines a DSL to configure GDK.

Most of the magic happens through the class method
`.method_missing`. The implementation of this method will dynamically
define instance methods for configuration settings.

Below is an example subclass of `GDK::ConfigSettings` to demonstrate
each kind.

```ruby
class ExampleConfig < GDK::ConfigSettings
  foo 'hello'
  bar { rand(1..10) }
  fuz do |f|
    f.buz 1234
  end
end
```

- `foo`: (literal value) This is just a literal value, it can be any
  type (e.g. Number, Boolean, String).
- `bar`: (block without argument) This is using a block to set a
  value. It evaluates the Ruby code to dynamically calculate a value.
- `fuz`: (block with argument) When the block takes a single argument,
  it expects you'll be setting child settings.

If you'd dump this configuration with `rake dump_config` you'll get something
like:

```yaml
foo: hello
bar: 5
fuz:
  buz: 1234
```

When you use a block without argument you can also calculate a value
based on another setting. So for example, we'd could replace the `bar`
block with `{ config.fuz.buz + 1000 }` and then the value would be
`2234`.

#### `GDK::Config`

`GDK::Config` is the single source of truth when it comes down to
defaults. In this file, every existing setting is specified and for
each setting a default is provided.

#### Dynamic settings

Some settings in `GDK::Config` are prepended with `__` (double
underscore). These are not supposed to be set in `gdk.yml` and only
act as a intermediate value. They also will not be shown by `#dump!`.

### Adding a setting

When you add a new setting:

1. Add it to `lib/gdk/config.rb`.
1. Run `rake gdk.example.yml` to regenerate this file.
1. Commit both files.

## Webpack Settings

### Webpack `gdk.yml` settings

Under the webpack key you can define the following settings with their defaults:

```yaml
webpack:
  host: 127.0.0.1
  port: 3808
  static: false
  vendor_dll: false
```

| Setting | Default | Description |
| --- | ------ | ----- |
| `host` | `127.0.0.1` | The host your webpack dev server is running on. Usually no need to change. |
| `port` | `3808` | The port your webpack dev server is running on. You should change this if you are running multiple GDKs |
| `static` | `false` | Setting this to `true` will replace the webpack dev server with a lightweight Ruby server. See below for more information |
| `vendor_dll` | `false` | Setting this to `true` will move certain dependencies to a webpack DLL. See below for more information |
| `sourcemaps` | `true` | Setting this to `false` will disable sourcemaps. This will reduce memory consumption for those who do not need to debug frontend code. |

#### Saving memory on the webpack dev server

By default webpack will run a dev server which watches file changes and keeps all the frontend assets in memory.
This allows for very quick recompile times, but is very memory demanding.
There are two settings from above which can help with this.

If you do not require frequent frontend asset recompiles, for example you are a backend
developer who rarely changes frontend files, you can enable `webpack.static: true` in
your `gdk.yml`. All frontend assets will be compiled once when GDK starts and
again, from scratch, if any frontend source or dependency file changes (for example,
branch switches).

This means you pay a high upfront cost of a single compile (a lot of memory and CPU),
but if you do not change any frontend files, you will just have a lightweight Ruby
server running.

The `webpack.vendor_dll: true` mode is an alternate memory saving mode, which takes
infrequently updated dependencies and combines them into one long-lived bundle that is
written to disk and does not reside in memory. We have seen around 200 to 300 MB in
memory savings.

This mode is recommended for everyone. It is still disabled by default, because we want
to get more real usage data. One of the side-effects could be an increasingly large
`gitlab/tmp` directory. We will eventually make this the default.

If you experience any problems with one of the modes, you can quickly change the settings in your `gdk.yml` and regenerate the Procfile.

### Webpack ENV variables

The GitLab application exposes various configuration options for webpack via
ENV variables. These can be modified to improve performance or enable debugging.

These settings can be configured using [`env.runit`](runit.md#modifying-environment-configuration-for-services).

| Variable | Default | Description |
| ------------- | ------- | ----------- |
| DEV_SERVER_LIVERELOAD | true | Disables live reloading of frontend assets |
| NO_COMPRESSION        | false | Disables compression of assets |
| NO_SOURCEMAPS         | false | Disables generation of sourcemaps (reduces size of `main.chunk.js` by ~50%) |
| WEBPACK_MEMORY_TEST   | false | Output the in-memory heap size upon compilation and exit |
| WEBPACK_CACHE_PATH    | `./tmp/cache` | Path string to temporary dir     |
| WEBPACK_REPORT        | false       | Generates bundle analysis report |
| WEBPACK_VENDOR_DLL    | false       | Reduce webpack-dev-server memory requirements when vendor bundle has been precompiled with `yarn webpack-vendor` |


CHANGELOG.md
# CHANGELOG

This Changelog tracks major changes to the GitLab Development Kit,
such as dependency updates (those not handled by Bundler) and new
features.

## 2018-01-11

- Added ChromeDriver to deprecate PhantomJS. !423 !380

## 2017-11-24

- Add Docker-in-Docker support in `Vagrantfile` (needed to run GitLab Container Registry)

## 2017-11-21

- [GitLab Geo] Add replication slot

## 2017-02-01

- Add webpack process to `Procfile` configure it within `gitlab.yml` !237
  Make sure to [update GDK](doc/update-gdk.md) and read the
  [troubleshooting section](doc/howto/troubleshooting.md#webpack).

## 2016-10-31

- Add root check to catch root move problems. Requires gem 0.2.3 or
  newer. Next time you run `git pull` in the `gitlab-development-kit`
  root directory, also run `gem install gitlab-development-kit` to
  upgrade the gem.

## 2016-09-09

- Update `Procfile` for `gitlab_workhorse_secret`

## 2016-09-05

- Added a Changelog.

## 2016-08-16

- Updated PhantomJS to 2.1.1. !182

## 2016-08-11

- Updated Ruby to 2.3.1. !178

## 2016-08-08

- Added the [`gitlab-development-kit` gem](https://rubygems.org/gems/gitlab-development-kit), commands can now be run using the `gdk` CLI. !174
- Began using a `GOPATH` for GitLab Workhorse, this change requires manual intervention. [See the update instructions here](https://gitlab.com/gitlab-org/gitlab-development-kit/blob/fd04b7f1a3a72302af71c1a7923daaa5b22dcd28/gitlab-workhorse/README.md#cleaning-up-an-old-gitlab-workhorse-checkout). !173


lib/gdk/config.rb
# frozen_string_literal: true

require 'etc'
require 'cgi'
require_relative 'config_settings'

module GDK
  class Config < ConfigSettings
    GDK_ROOT = Pathname.new(__dir__).parent.parent
    FILE = File.join(GDK_ROOT, 'gdk.yml')

    settings :repositories do
      string(:gitlab) { 'https://gitlab.com/gitlab-org/gitlab.git' }
      string(:gitlab_shell) { 'https://gitlab.com/gitlab-org/gitlab-shell.git' }
      string(:gitlab_workhorse) { 'https://gitlab.com/gitlab-org/gitlab-workhorse.git' }
      string(:gitaly) { 'https://gitlab.com/gitlab-org/gitaly.git' }
      string(:gitlab_pages) { 'https://gitlab.com/gitlab-org/gitlab-pages.git' }
      string(:gitlab_docs) { 'https://gitlab.com/gitlab-com/gitlab-docs.git' }
      string(:gitlab_elasticsearch_indexer) { 'https://gitlab.com/gitlab-org/gitlab-elasticsearch-indexer.git' }
    end

    array(:git_repositories) do
      # This list in not exhaustive yet, as some git repositories are based on
      # a fake GOPATH inside a projects sub directory
      %w[/ gitlab]
        .map { |d| File.join(gdk_root, d) }
        .select { |d| Dir.exist?(d) }
    end

    path(:gdk_root) { self.class::GDK_ROOT }

    settings :gdk do
      bool(:ask_to_restart_after_update) { true }
      bool(:debug) { false }
      settings :experimental do
        bool(:quiet) { false }
        bool(:auto_reconfigure) { false }
      end
      bool(:overwrite_changes) { false }
      array(:protected_config_files) { [] }
    end

    path(:repositories_root) { config.gdk_root.join('repositories') }

    string(:listen_address) { '127.0.0.1' }

    string :hostname do
      next "#{config.auto_devops.gitlab.port}.qa-tunnel.gitlab.info" if config.auto_devops?

      read!('hostname') || read!('host') || config.listen_address
    end

    integer :port do
      next 443 if config.auto_devops?

      read!('port') || 3000
    end

    settings :https do
      bool :enabled do
        next true if config.auto_devops?

        read!('https_enabled') || false
      end
    end

    string :relative_url_root do
      read!('relative_url_root') || '/'
    end

    anything :__uri do
      # Only include the port if it's 'non standard'
      klass = config.https? ? URI::HTTPS : URI::HTTP
      relative_url_root = config.relative_url_root.gsub(%r{\/+$}, '')

      klass.build(host: config.hostname, port: config.port, path: relative_url_root)
    end

    string(:username) { Etc.getpwuid.name }

    settings :webpack do
      string :host do
        next config.auto_devops.listen_address if config.auto_devops?

        read!('webpack_host') || config.hostname
      end
      bool(:static) { false }
      bool(:vendor_dll) { false }
      bool(:sourcemaps) { true }

      integer(:port) { read!('webpack_port') || 3808 }
    end

    settings :workhorse do
      integer(:configured_port) { 3333 }

      string :__active_host do
        next config.auto_devops.listen_address if config.auto_devops?

        config.hostname
      end

      integer :__active_port do
        if config.auto_devops? || config.nginx?
          config.workhorse.configured_port
        else
          # Workhorse is the user-facing entry point whenever neither nginx nor
          # AutoDevOps is used, so in that situation use the configured GDK port.
          config.port
        end
      end

      bool(:auto_update) { true }
    end

    settings :gitlab_shell do
      bool(:auto_update) { true }
      string(:dir) { config.gdk_root.join('gitlab-shell') }
    end

    settings :gitlab_elasticsearch_indexer do
      bool(:auto_update) { true }
      path(:__dir) { config.gdk_root.join('gitlab-elasticsearch-indexer') }
    end

    settings :registry do
      bool :enabled do
        next true if config.auto_devops?

        read!('registry_enabled') || false
      end

      string :host do
        next "#{config.auto_devops.registry.port}.qa-tunnel.gitlab.info" if config.auto_devops?

        config.hostname
      end

      string :api_host do
        next config.listen_address if config.auto_devops?

        config.hostname
      end

      string :tunnel_host do
        next config.listen_address if config.auto_devops?

        config.hostname
      end

      integer(:tunnel_port) { 5000 }

      integer :port do
        read!('registry_port') || 5000
      end

      string :image do
        read!('registry_image') ||
          'registry.gitlab.com/gitlab-org/build/cng/gitlab-container-registry:'\
        'v2.9.1-gitlab'
      end

      integer :external_port do
        next 443 if config.auto_devops?

        5000
      end

      bool(:self_signed) { false }
      bool(:auth_enabled) { true }
    end

    settings :object_store do
      bool(:enabled) { read!('object_store_enabled') || false }
      string(:host) { config.listen_address }
      integer(:port) { read!('object_store_port') || 9000 }
    end

    settings :gitlab_pages do
      bool(:enabled) { true }
      integer(:port) { read!('gitlab_pages_port') || 3010 }
      bool(:auto_update) { true }
    end

    settings :auto_devops do
      bool(:enabled) { read!('auto_devops_enabled') || false }
      string(:listen_address) { '0.0.0.0' }
      settings :gitlab do
        integer(:port) { read_or_write!('auto_devops_gitlab_port', rand(20_000..24_999)) }
      end
      settings :registry do
        integer(:port) { read!('auto_devops_registry_port') || (config.auto_devops.gitlab.port + 5000) }
      end
    end

    settings :omniauth do
      settings :google_oauth2 do
        string(:enabled) { !!read!('google_oauth_client_secret') || '' }
        string(:client_id) { read!('google_oauth_client_id') || '' }
        string(:client_secret) { read!('google_oauth_client_secret') || '' }
      end
    end

    settings :geo do
      bool(:enabled) { false }
      string(:node_name) { config.gdk_root.basename.to_s }
      settings :registry_replication do
        bool(:enabled) { false }
        string(:primary_api_url) { 'http://localhost:5000' }
      end
    end

    settings :elasticsearch do
      bool(:enabled) { false }
      string(:version) { '7.7.0' }
      string(:mac_checksum) { '189e207d7c95f831df2973680fe7a1ebc5ade83705c3af38befc23a9e8c0893636b9d9699a153e69b6abaa3c5cdb1bb8e459596cb4e31115afe5a88f48295b61' }
      string(:linux_checksum) { 'bc1b0409128288b9eb1e71d2be0de432016d75aec3fc9ec9b6e8b3dc7eb9d4fd0012fe05b0ee81e2ddfa4be005962bae9f64d01dc7f4814eb767943f16ef0dfb' }
    end

    settings :tracer do
      string(:build_tags) { 'tracer_static tracer_static_jaeger' }
      settings :jaeger do
        bool(:enabled) { true }
        string(:version) { '1.10.1' }
      end
    end

    settings :nginx do
      bool(:enabled) { false }
      string(:listen) { config.hostname }
      string(:bin) { find_executable!('nginx') || '/usr/local/bin/nginx' }
      settings :ssl do
        string(:certificate) { 'localhost.crt' }
        string(:key) { 'localhost.key' }
      end
      settings :http do
        bool(:enabled) { false }
        integer(:port) { 8080 }
      end
      settings :http2 do
        bool(:enabled) { false }
      end
    end

    settings :postgresql do
      integer(:port) { read!('postgresql_port') || 5432 }
      path(:bin_dir) { cmd!(%w[support/pg_bindir]) || '/usr/local/bin' }
      path(:bin) { config.postgresql.bin_dir.join('postgres') }
      string(:replication_user) { 'gitlab_replication' }
      path(:dir) { config.gdk_root.join('postgresql') }
      path(:data_dir) { config.postgresql.dir.join('data') }
      string(:host) { config.postgresql.dir.to_s }
      path(:replica_dir) { config.gdk_root.join('postgresql-replica') }
      settings :replica do
        bool(:enabled) { false }
      end
      settings :geo do
        integer(:port) { 5431 }
        path(:dir) { config.gdk_root.join('postgresql-geo') }
        string(:host) { config.postgresql.geo.dir.to_s }
      end
    end

    settings :gitaly do
      path(:address) { config.gdk_root.join('gitaly.socket') }
      path(:assembly_dir) { config.gdk_root.join('gitaly', 'assembly') }
      path(:config_file) { config.gdk_root.join('gitaly', 'gitaly.config.toml') }
      path(:internal_socket_dir) { config.gdk_root.join('tmp', 'gitaly') }
      path(:log_dir) { config.gdk_root.join('log', 'gitaly') }
      bool(:auto_update) { true }
    end

    settings :praefect do
      path(:address) { config.gdk_root.join('praefect.socket') }
      path(:config_file) { config.gdk_root.join("gitaly", "praefect.config.toml") }
      bool(:enabled) { true }
      path(:internal_socket_dir) { config.gdk_root.join('tmp', 'praefect') }
      settings :database do
        path(:host) { config.postgresql.dir }
        string(:dbname) { 'praefect_development' }
        string(:sslmode) { 'disable' }
      end
      integer(:node_count) { 1 }
      array(:nodes) do
        settings_array!(config.praefect.node_count) do |i|
          path(:address) { config.gdk_root.join("gitaly-praefect-#{i}.socket") }
          string(:config_file) { "gitaly/gitaly-#{i}.praefect.toml" }
          path(:log_dir) { config.gdk_root.join("log", "praefect-gitaly-#{i}") }
          bool(:primary) { i.zero? }
          string(:service_name) { "praefect-gitaly-#{i}" }
          string(:storage) { "praefect-internal-#{i}" }
          path(:storage_dir) { i.zero? ? config.repositories_root : File.join(config.repositories_root, storage) }
          path(:internal_socket_dir) { config.gdk_root.join('tmp', 'praefect', "gitaly-#{i}") }
        end
      end
    end

    settings :sshd do
      bool(:enabled) { false }
      path(:bin) { find_executable!('sshd') || '/usr/local/sbin/sshd' }
      string(:listen_address) { config.hostname }
      integer(:listen_port) { 2222 }
      string(:user) { config.username }
      path(:authorized_keys_file) { config.gdk_root.join('.ssh', 'authorized_keys') }
      path(:host_key) { config.gdk_root.join('openssh', 'ssh_host_rsa_key') }
      string(:additional_config) { '' }
    end

    settings :git do
      path(:bin) { find_executable!('git') || '/usr/local/bin/git' }
    end

    settings :runner do
      path(:config_file) { config.gdk_root.join('gitlab-runner-config.toml') }
      bool(:enabled) { !!read!(config.runner.config_file) }
      array(:extra_hosts) { [] }
      string(:token) { 'DEFAULT TOKEN: Register your runner to get a valid token' }
    end

    settings :grafana do
      bool(:enabled) { false }
    end

    settings :prometheus do
      bool(:enabled) { false }
    end

    settings :openldap do
      bool(:enabled) { false }
    end

    settings :mattermost do
      bool(:enabled) { false }
      integer(:port) { config.auto_devops.gitlab.port + 7000 }
      string(:image) { 'mattermost/mattermost-preview' }
      integer(:local_port) { 8065 }
    end

    settings :gitlab do
      path(:dir) { config.gdk_root.join('gitlab') }
      path(:__socket_file) { config.gdk_root.join('gitlab.socket') }
      string(:__socket_file_escaped) { CGI.escape(config.gitlab.__socket_file.to_s) }

      settings :actioncable do
        path(:__socket_file) { config.gdk_root.join('gitlab.actioncable.socket') }
      end
    end
  end
end
